File: app/schemas/advance_repasse_ledger.py
from __future__ import annotations

from datetime import date, datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.common import ORMModel


class RepasseLedgerEntryRead(ORMModel):
    id: UUID
    institution_id: UUID
    direction: str  # CREDIT | DEBIT
    amount: float
    source_type: str  # OPERATION | SETTLEMENT | WITHDRAWAL | ADJUSTMENT
    withdrawal_purpose: str | None = None  # DEBT_REDUCTION | OTHER (só em WITHDRAWAL)
    #: Dívida abatida por esta retirada (só em DEBT_REDUCTION). Vira pagamento na Evolução.
    debt_item_id: UUID | None = None
    source_batch_id: UUID | None = None
    source_movement_id: UUID | None = None
    occurred_at: date
    description: str | None = None
    reversed_at: datetime | None = None
    reversal_reason: str | None = None
    created_at: datetime


class RepasseLedgerStatementRead(BaseModel):
    institution_id: UUID | None = None
    balance: float
    entries: list[RepasseLedgerEntryRead] = Field(default_factory=list)


class RepasseWithdrawalCreate(BaseModel):
    """Entrada da Retirada de Repasse — destino ESTRUTURADO; descrição é só observação livre."""

    institution_id: UUID
    amount: float = Field(gt=0)
    occurred_at: date
    purpose: Literal["DEBT_REDUCTION", "OTHER"]
    #: Qual dívida a retirada abate. Só faz sentido com purpose=DEBT_REDUCTION; sem ele a
    #: retirada apenas reduz o saldo do repasse, sem virar pagamento de dívida nenhuma.
    debt_item_id: UUID | None = None
    description: str | None = None


class RepasseDebtLinkIn(BaseModel):
    """Vincula/desvincula uma retirada à dívida que ela abate. `null` desvincula."""

    debt_item_id: UUID | None = None

File: frontend/src/services/advanceRepasseLedger.ts
import { api } from "./api";

// Extrato de Repasse (append-only). O usuário vê apenas "Entradas/Saídas" e o "Saldo" — nunca
// precisa entender Ledger. Tudo vem pronto do backend.

export type LedgerDirection = "CREDIT" | "DEBIT";
export type LedgerSourceType = "OPERATION" | "SETTLEMENT" | "WITHDRAWAL" | "ADJUSTMENT";
export type WithdrawalPurpose = "DEBT_REDUCTION" | "OTHER";

export const DIRECTION_LABELS: Record<LedgerDirection, string> = {
  CREDIT: "Entrada",
  DEBIT: "Saída",
};

export const SOURCE_LABELS: Record<LedgerSourceType, string> = {
  OPERATION: "Operação de antecipação",
  SETTLEMENT: "Liquidação de NF",
  WITHDRAWAL: "Retirada de Repasse",
  ADJUSTMENT: "Ajuste",
};

// Destino estruturado da retirada — rótulo próprio (nunca dentro da descrição livre).
export const WITHDRAWAL_PURPOSE_LABELS: Record<WithdrawalPurpose, string> = {
  DEBT_REDUCTION: "Abatimento de dívida",
  OTHER: "Outros",
};

export interface LedgerEntry {
  id: string;
  institution_id: string;
  direction: LedgerDirection;
  amount: number;
  source_type: LedgerSourceType;
  withdrawal_purpose: WithdrawalPurpose | null;
  source_batch_id: string | null;
  source_movement_id: string | null;
  occurred_at: string;
  description: string | null;
  reversed_at: string | null;
  reversal_reason: string | null;
  created_at: string;
}

export interface LedgerStatement {
  institution_id: string | null;
  balance: number;
  entries: LedgerEntry[];
}

export async function fetchRepasseLedger(
  institutionId?: string | null,
  includeReversed = true,
): Promise<LedgerStatement> {
  const params: Record<string, string | boolean> = { include_reversed: includeReversed };
  if (institutionId) params.institution_id = institutionId;
  const { data } = await api.get<LedgerStatement>("/invoices/advance-repasse-ledger", { params });
  return data;
}

export interface WithdrawalInput {
  institution_id: string;
  amount: number;
  occurred_at: string; // YYYY-MM-DD
  purpose: WithdrawalPurpose;
  description?: string | null;
}

/** Registra uma Retirada de Repasse (DÉBITO append-only). Reduz apenas o saldo do Repasse. */
export async function createRepasseWithdrawal(input: WithdrawalInput): Promise<LedgerEntry> {
  const { data } = await api.post<LedgerEntry>(
    "/invoices/advance-repasse-ledger/withdrawals",
    input,
  );
  return data;
}
